src/main/java/net/neoforged/bus/ListenerList.java

```diff
@@ -30,247 +29,109 @@
-    }
-
-    static void resize(int max)
-    {
-        if (max > maxSize)
-        {
-            synchronized (ListenerList.class)
-            {
-                if (max > maxSize)
-                {
-                    allLists.forEach(list -> list.resizeLists(max));
-                    maxSize = max;
-                }
-            }
+        for (int x = 0; x < count; x++) {
+            priorities.add(new ArrayList<>());
         }
     }
 
-    private synchronized void resizeLists(int max)
-    {
-        if (parent != null)
-        {
-            parent.resizeLists(max);
-        }
-
-        if (lists.length >= max)
-        {
-            return;
-        }
 
-        ListenerListInst[] newList = new ListenerListInst[max];
-        int x = 0;
-        for (; x < lists.length; x++)
-        {
-            newList[x] = lists[x];
-        }
-        for(; x < max; x++)
-        {
-            if (parent != null)
-            {
-                newList[x] = new ListenerListInst(parent.getInstance(x));
-            }
-            else
-            {
-                newList[x] = new ListenerListInst();
-            }
-        }
-        lists = newList;
+    ListenerList(ListenerList parent) {
+        this();
+        this.parent = parent;
+        this.parent.addChild(this);
     }
 
-    public static synchronized void clearBusID(int id)
-    {
-        for (ListenerList list : allLists)
-        {
-            list.lists[id].dispose();
-        }
+    /**
+     * Returns a ArrayList containing all listeners for this event,
+     * and all parent events for the specified priority.
+     * <p>
+     * The list is returned with the listeners for the children events first.
+     *
+     * @param priority The Priority to get
+     * @return ArrayList containing listeners
+     */
+    public ArrayList<IEventListener> getListeners(EventPriority priority) {
+        writeLock.acquireUninterruptibly();
+        ArrayList<IEventListener> ret = new ArrayList<>(priorities.get(priority.ordinal()));
+        writeLock.release();
+        if (parent != null) {
+            ret.addAll(parent.getListeners(priority));
+        }
+        return ret;
     }
 
-    protected ListenerListInst getInstance(int id)
-    {
-        return lists[id];
+    /**
+     * Returns a full list of all listeners for all priority levels.
+     * Including all parent listeners.
+     * <p>
+     * List is returned in proper priority order.
+     * <p>
+     * Automatically rebuilds the internal Array cache if its information is out of date.
+     *
+     * @return Array containing listeners
+     */
+    public IEventListener[] getListeners() {
+        if (shouldRebuild()) buildCache();
+        return listeners.get();
     }
 
-    public IEventListener[] getListeners(int id)
-    {
-        return lists[id].getListeners();
+    protected boolean shouldRebuild() {
+        return rebuild;// || (parent != null && parent.shouldRebuild());
```

Comment:
Why is this code commented out? If it's intentional, you should remove the comment.

Reply:
That was already the case before, I just moved the code.